bolt-modules/boltlib/lib/puppet/functions/run_task_with.rb

```diff
@@ -0,0 +1,179 @@
+  #     { 'param1' => $t.vars['var1'],
+  #       'param2' => $t.vars['var2'] }
+  #   }
+  dispatch :run_task_with_with_description do
+    param 'String[1]', :task_name
+    param 'Boltlib::TargetSpec', :targets
+    param 'Optional[String]', :description
+    optional_param 'Hash[String[1], Any]', :options
+    required_block_param 'Callable[Target]', :block
+    return_type 'ResultSet'
+  end
+
+  def run_task_with(task_name, targets, options = {}, &block)
+    run_task_with_with_description(task_name, targets, nil, options, &block)
+  end
+
+  def run_task_with_with_description(task_name, targets, description, options = {})
+    unless Puppet[:tasks]
+      raise Puppet::ParseErrorWithIssue
+        .from_issue_and_stack(
+          Bolt::PAL::Issues::PLAN_OPERATION_NOT_SUPPORTED_WHEN_COMPILING,
+          action: 'run_task_with'
+        )
+    end
+
+    executor  = Puppet.lookup(:bolt_executor)
+    inventory = Puppet.lookup(:bolt_inventory)
+
+    # Report to analytics
+    executor.report_function_call(self.class.name)
+    executor.report_bundled_content('Task', task_name)
+
+    # Keep valid metaparameters, discarding everything else
+    options = options.select { |k, _v| k.start_with?('_') }
+                     .transform_keys { |k| k.sub(/^_/, '').to_sym }
+
+    options[:description] = description if description
+
+    # Get all the targets
+    targets = Array(inventory.get_targets(targets))
+
+    # Map the targets to their specific parameters
+    target_mapping = targets.each_with_object({}) do |target, mapping|
+      params = yield(target)
+
+      unless params.is_a?(Hash)
+        raise with_stack(
+          :TYPE_MISMATCH,
+          "Block must return a Hash of parameters, received #{params.class} for target #{target.safe_name}"
+        )
+      end
+
+      mapping[target] = params
+    end
+
+    # If all targets use the 'pcp' transport, use a fake task instead of loading the local definition
+    # Otherwise, load the local task definition
+    if targets.any? && targets.all? { |t| t.transport == 'pcp' }
+      task = Bolt::Task.new(task_name, {}, [{ 'name' => '', 'path' => '' }])
+    else
+      task_signature = Puppet::Pal::ScriptCompiler.new(closure_scope.compiler).task_signature(task_name)
+
+      if task_signature.nil?
+        raise Bolt::Error.unknown_task(task_name)
+      end
+
+      task = Bolt::Task.from_task_signature(task_signature)
+
+      # Retrieve the default parameters for the task and merge them with the
+      # target-specific parameters
+      default_params = task.parameter_defaults
+      target_mapping.transform_values! { |params| default_params.merge(params) }
+
+      target_mapping.each_value do |params|
+        task_signature.runnable_with?(params) do |mismatch_message|
+          raise with_stack(:TYPE_MISMATCH, mismatch_message)
+        end || (raise with_stack(:TYPE_MISMATCH, 'Task parameters do not match'))
```

Comment:
This will abort the entire function if any target has invalid params, right? Is that the correct behavior or should it still try to run on targets that have valid params?

Reply:
Correct.

Would the following approach be more correct?
- Partition targets based on valid/invalid params
- Run tasks on targets with valid params
- Create failing results for targets with invalid params and add them to the resultset

Reply:
Updated to create failing results using `Bolt::Result.from_exception` and add them to the resultset returned from the task run.